beacon_node/beacon_chain/src/canonical_head.rs

```diff
@@ -0,0 +1,1307 @@
+                // The async task completed successfully.
+                Ok(Some(())) => Ok(()),
+                // The async task did not complete successfully since the runtime is shutting down.
+                Ok(None) => {
+                    debug!(
+                        self.log,
+                        "Did not update EL fork choice";
+                        "info" => "shutting down"
+                    );
+                    Err(Error::RuntimeShutdown)
+                }
+                // The async task did not complete successfully, tokio returned an error.
+                Err(e) => {
+                    error!(
+                        self.log,
+                        "Did not update EL fork choice";
+                        "error" => ?e
+                    );
+                    Err(Error::TokioJoin(e))
+                }
+            },
+            // There was an error recomputing the head.
+            Err(e) => {
+                metrics::inc_counter(&metrics::FORK_CHOICE_ERRORS);
+                Err(e)
+            }
+        }
+    }
+
+    /// A non-async (blocking) function which recomputes the canonical head and spawns async tasks.
+    ///
+    /// This function performs long-running, heavy-lifting tasks which should not be performed on
+    /// the core `tokio` executor.
+    fn recompute_head_at_slot_internal(
+        self: &Arc<Self>,
+        current_slot: Slot,
+    ) -> Result<Option<JoinHandle<Option<()>>>, Error> {
+        let recompute_head_lock = self.canonical_head.recompute_head_lock.lock();
+
+        // Take a clone of the current ("old") head.
+        let old_cached_head = self.canonical_head.cached_head();
+
+        // Determine the current ("old") fork choice parameters.
+        //
+        // It is important to read the `fork_choice_view` from the cached head rather than from fork
+        // choice, since the fork choice value might have changed between calls to this function. We
+        // are interested in the changes since we last cached the head values, not since fork choice
+        // was last run.
+        let old_view = ForkChoiceView {
+            head_block_root: old_cached_head.head_block_root(),
+            justified_checkpoint: old_cached_head.justified_checkpoint(),
+            finalized_checkpoint: old_cached_head.finalized_checkpoint(),
+        };
+
+        let mut fork_choice_write_lock = self.canonical_head.fork_choice_write_lock();
+
+        // Recompute the current head via the fork choice algorithm.
+        fork_choice_write_lock.get_head(current_slot, &self.spec)?;
+
+        // Read the current head value from the fork choice algorithm.
+        let new_view = fork_choice_write_lock.cached_fork_choice_view();
+
+        // Downgrade the fork choice write-lock to a read lock, without allowing access to any
+        // other writers.
+        let fork_choice_read_lock = RwLockWriteGuard::downgrade(fork_choice_write_lock);
```

Comment:
Looks like you could reduce lock contention here by moving this lock downgrade directly after `fork_choice_write_lock.get_head(current_slot, &self.spec)` - I *think* the `cached_fork_choice_view()` call causes a read from disk?

```suggestion
        // Downgrade the fork choice write-lock to a read lock, without allowing access to any
        // other writers.
        let fork_choice_read_lock = RwLockWriteGuard::downgrade(fork_choice_write_lock);

        // Read the current head value from the fork choice algorithm.
        let new_view = fork_choice_read_lock.cached_fork_choice_view();
```

Reply:
Nice one! `cached_fork_choice_view` doesn't read from disk, but it's nice to hold the write-lock as little as possible. I've addressed this :pray: